Improve Past Groups table on Contact Groups tab

larssandergreen opened the pull request:
Overview
----------------------------------------
The Past groups table on the Contact Groups tab is somewhat confusing and sometimes provides unexpected results for Date Added or Date Removed. In general this table uses concepts that encourage misunderstandings of what it means for a contact to be removed from a group, doesn't differentiate between smart and regular groups even though these behave very differently here, and possibly leads to users doing something they don't intend to do (putting a contact back in a smart group by deleting them from a group, for example).

While group memberships concepts remain confusing, I hope these changes clarify the situation somewhat for users.

Before
----------------------------------------
![image](https://user-images.githubusercontent.com/25517556/194783150-7a36818b-b6c7-4c00-958d-072fa2b5d95d.png)

<img width="500" alt="Screen Shot 2022-10-09 at 4 46 29 PM" src="https://user-images.githubusercontent.com/25517556/194783182-67bd2aa8-12e4-4818-be28-2ad65e690983.png">

<img width="650" alt="Screen Shot 2022-10-09 at 4 46 29 PM" src="https://user-images.githubusercontent.com/25517556/194783167-88357827-1b59-48aa-81c0-bcc4befa674c.png">

After
----------------------------------------
<img width="1008" alt="Screen Shot 2022-10-09 at 4 10 39 PM" src="https://user-images.githubusercontent.com/25517556/194783252-d0b5f92c-5c01-49c5-a6a7-7e1209139202.png">

Changed heading to Removed Groups (some of the groups may not in fact be past groups, because a contact can be removed from a group they were never in). I think it's important not to encourage that misunderstanding (I know I misunderstood this for years).
Added * to indicate smart groups.
Changed "Rejoin Group" text for smart groups to "Manual Add" to indicate that this action has different results.

The less obvious change here is what the Date Added and Date Removed are in special cases.
- If a contact has been removed more than once from a group since the most recent date they were added or deleted, the Date Removed is the earliest date after they were added or deleted. If they were never added or deleted, the date is the earliest removed date. In other words, we show the date that the contact was effectively removed from the group, not the most recent date their status was set to removed. I think this is much more in line with what removed intuitively means and I can't see much reason you'd want to know the date the contact was most recently removed again instead.
- If a contact has been deleted from a group and then removed after that, they no longer show a Date Added (because at the time of removal, they were not in the group). 
 
At some point, I plan to add a History link that will show the complete Subscription History for each group.

Added help text and Actions header to attach it to.
<img width="366" alt="Screen Shot 2022-10-09 at 4 10 47 PM" src="https://user-images.githubusercontent.com/25517556/194783377-4e65051c-b4dd-4bdb-86da-01c34806ecab.png">

For smart groups, changed the confirmation popup text:
<img width="500" alt="Screen Shot 2022-10-09 at 4 46 29 PM" src="https://user-images.githubusercontent.com/25517556/194783546-4ffe9a4b-aafe-4298-87d7-8ae05da5898c.png">

<img width="650" alt="Screen Shot 2022-10-09 at 4 46 42 PM" src="https://user-images.githubusercontent.com/25517556/194784691-09cfef69-891f-41fc-a4bc-6b337b1179f1.png">
Confirmation popups for regular groups are the same, except the delete one says "This group will no longer be listed under Removed Groups."

Unfortunately, this adds a number of translation strings, but I think the added clarity is worth it. I've tried to minimize this where possible.

larssandergreen committed:
Improve Past Groups table

seamuslee001 commented on templates/CRM/Contact/Page/View/GroupContact.tpl:
```diff
@@ -125,24 +125,39 @@
-                {ts}Delete{/ts}</a>{/if}
+            <td>
+              {if $permission EQ 'edit'}
+              {if $row.saved_search_id}
```
@larssandergreen are you able to indent this if and the If at 153 as well just so it is clearer how these ifs are structured

larssandergreen replied:
Done and simplified it a little, hopefully.

larssandergreen committed:
clarify nested ifs

seamuslee001 commented on templates/CRM/Contact/Page/View/GroupContact.tpl:
```diff
@@ -125,24 +125,42 @@
-                  {ts}Rejoin Group{/ts}</a>
-              <a class="action-item crm-hover-button" href="#Deleted" title="{ts 1=$displayName 2=$row.title}Delete %1 from %2? (this group will no longer be listed under Past Groups).{/ts}">
-                {ts}Delete{/ts}</a>{/if}
+                  {ts}Rejoin Group{/ts}
```
This needs a </a> on the end (note you open the a tag on New Line 148)

larssandergreen replied:
Yes, this was poorly done, sorry. Now should be better.

seamuslee001 commented on templates/CRM/Contact/Page/View/GroupContact.tpl:
```diff
@@ -125,24 +125,42 @@
+              {if $permission EQ 'edit'}
+                {if $row.saved_search_id}
+                <a class="action-item crm-hover-button" href="#Added" title="{ts 1=$displayName 2=$row.title}Add %1 manually into %2, overriding smart group critera?{/ts}">
+                  {ts}Manual Add{/ts}
```
this needs a </a> tag

larssandergreen committed:
cleanup links and ifs

aydun commented:
These look like useful improvements, but maybe the effort would be better directed into creating a SearchKit replacement for the Groups tab?

larssandergreen commented:
@aydun Not sure I could come up with a search that would replicate the Removed Groups table. But if you can?

colemanw commented:
Like this: 
![image](https://user-images.githubusercontent.com/2874912/195911009-a106a549-c739-4b13-a12b-0a3293e7deb3.png)

larssandergreen commented:
@colemanw For sure, but the part that's not obvious to me is how we get the Date Added and Date Removed (and specifically the dates as discussed in After above).

larssandergreen commented:
Yes, I think the thing that would be necessary would be being able to join on Contact_SubscriptionHistory_contact_id_01.date > (a different select) — and I can't see a way to do that with SK.

E.g. we want to find the earliest removed status row in subscription history that is later than the most recent added or deleted row and display that as the removed date.

civicrm-builder commented:
Can one of the admins verify this patch?

larssandergreen commented:
jenkins test this please

colemanw commented:
@larssandergreen sorry this PR has been in purgatory so long. I think the combination of 1) low-priority 2) confusing and 3) maybe-replaceable-by-SK have caused it to languish.

I was trying to understand your logic for which added and removed dates to show and wondering if I agreed with it, and then it occurred to me that maybe we *don't* have to pick a single date. By default when you do the join I illustrated in my previous comment and use grouping, then SK will aggregate using "list" which shows them all. Then you could see at a glance *all* the dates a contact was added and *all* the dates they were removed. That could actually be useful information.

So if we did it that way, then I *think* (haven't tried but I think) the table *can* be replaced with SK.

larssandergreen commented:
@colemanw No worries. I think we do want to have a single date for these as a list would be too confusing for general users, who just want to know when the contact was removed from the group in some kind of sane way. Part of the problem is that contacts can be removed from groups that they are not currently members of, so I don't think we actually want to show those removal dates - we just want the date that the contact went from being a member of the group to not being a member of the group.

My plan is actually to add a little pop-up that shows the subscription history (which would of course use SK), so you can actually see the full list of added, removed and deleted in order, which is kind of the same info you are talking about, but I think in a more useful format.

colemanw commented:
Ok, in that case I think the best you could do with SK is to show the most recent added and removal dates, and conditionally hide the removal date if there is no added date.
I think that's pretty much what you want, right?

larssandergreen commented:
What I've done here is to show the first removal date after the most recent added date. So if you have:

Added Jan 1
Removed Feb 1
Added Jun 1
Removed Sept 1
Removed Dec  1

The proposed table will show Added June 1, Removed Sept 1 (instead of Added Jan 1, Removed Dec 1).

I think that's much more helpful and more in line with what you would expect than the most recent removal date. For example, let's say I'm looking into why someone did or didn't receive our email newsletter and the current past groups table in this example says the contact was removed on Dec 1, so why didn't they get our Sept, Oct and Nov newsletters? Weird. Well, they were actually removed Sept 1, but that fact isn't apparent.

We wouldn't want to hide the removed data if there is no added date, as this is relevant for smart groups.

colemanw commented:
Ok I'm gonna go ahead and merge this since we don't have a ready alternative in SK.

larssandergreen commented:
Thanks Coleman!

On Wed, Jun 7, 2023 at 4:28 PM colemanw ***@***.***> wrote:

> Merged #24705 <https://github.com/civicrm/civicrm-core/pull/24705> into
> master.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/civicrm/civicrm-core/pull/24705#event-9464905435>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGCV35A6UJE6M6JMLQGBVMDXKD57LANCNFSM6AAAAAARA4OWTU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
